src/compose/MentionWarnings.js

```diff
@@ -0,0 +1,140 @@
+
+class MentionWarnings extends PureComponent<Props, State> {
+  state = {
+    unsubscribedMentions: [],
+  };
+
+  /**
+   * Check whether the message text entered by the user contains
+   * an @-mention to a user unsubscribed to the current stream, and if
+   * so, shows a warning.
+   *
+   * This function is expected to be called by `ComposeBox` using a ref
+   * to this component.
+   *
+   * @param message The message text entered by the user in the compose box.
+   */
+  handleMentionSubscribedCheck = async (message: string) => {
+    const { usersById, narrow, auth, stream } = this.props;
+
+    if (isPrivateNarrow(narrow)) {
+      return;
+    }
+    const unformattedMessage = message.split('**')[1];
+
+    // We skip user groups, for which autocompletes are of the form
+    // `*<user_group_name>*`, and therefore, message.split('**')[1]
+    // is undefined.
+    if (unformattedMessage === undefined) {
+      return;
+    }
+    const [userFullName, userId] = unformattedMessage.split('|');
+    const unsubscribedMentions = this.state.unsubscribedMentions.slice();
+    let mentionedUser: UserOrBot;
+
+    // eslint-disable-next-line no-unused-vars
+    for (const [id, user] of usersById) {
+      if (userId !== undefined) {
+        if (user.user_id === userId) {
+          mentionedUser = user;
+          break;
+        }
+      } else if (user.full_name === userFullName) {
+        mentionedUser = user;
+        break;
+      }
+    }
+    if (!mentionedUser || unsubscribedMentions.includes(mentionedUser)) {
+      return;
+    }
+
+    if (
+      !(await api.getSubscriptionToStream(auth, mentionedUser.user_id, stream.stream_id))
+        .is_subscribed
+    ) {
+      unsubscribedMentions.push(mentionedUser);
+      this.setState({ unsubscribedMentions });
+    }
+  };
+
+  handleMentionWarningDismiss = (user: UserOrBot) => {
+    this.setState(prevState => ({
+      unsubscribedMentions: prevState.unsubscribedMentions.filter(
+        (x: UserOrBot) => x.user_id !== user.user_id,
+      ),
+    }));
+  };
+
+  clearMentionWarnings = () => {
+    this.setState({
+      unsubscribedMentions: [],
+    });
+  };
+
+  render() {
+    const { unsubscribedMentions } = this.state;
+    const { stream, narrow } = this.props;
+
+    if (isPrivateNarrow(narrow)) {
+      return null;
+    }
+
+    const mentionWarnings = [];
+    for (const user of unsubscribedMentions) {
+      mentionWarnings.push(
+        <MentionedUserNotSubscribed
+          stream={stream}
+          user={user}
+          onDismiss={this.handleMentionWarningDismiss}
+          key={user.user_id}
+        />,
+      );
+    }
+
+    return (
+      <AnimatedScaleComponent visible={mentionWarnings.length !== 0}>
+        {mentionWarnings}
+      </AnimatedScaleComponent>
+    );
+  }
+}
+
+// TODO: Use a type checked connect call.  $FlowFixMe.
+export default connect(
+  (state, props) => ({
+    auth: getAuth(state),
+    usersById: getActiveUsersById(state),
+  }),
```

Comment:
Better to leave out the `props` parameter here when it's not going to be used. React Redux inspects this callback to see if it takes a second parameter, and if it doesn't then it applies an optimization where it knows it doesn't need to re-evaluate anything if the props change (and instead only if the state changes.)